src/core.js

```diff
@@ -101,52 +124,105 @@ async function pruneNonCriticalCssLauncher ({
+                await page.waitFor(500)
+                page
+                  .evaluate(pageLoadSkipTimeoutFunc, {
+                    pageLoadSkipTimeout
+                  })
+                  .then(message => {
+                    if (!_hasLoaded) {
+                      // when pageLoadSkipTimeout is reached after dom request
+                      // don't resolve when page already loaded completely
+                      debuglog(
+                        'pageLoadSkipTimeout - page load waiting ABORTED after ' +
+                          pageLoadSkipTimeout / 1000 +
+                          's. '
+                      )
+                      return resolve('pageLoadSkipTimeout')
+                    }
+                  })
+                  .catch(err => {
+                    if (!err.message.includes('Target closed')) {
+                      debuglog('page.evaluate - ERROR', err)
+                      _hasError = err
+                      return reject(err)
+                    }
+                  })
+              }
+            }
+          })
         }
       })
 
       debuglog('page load start')
       // set a higher number than the timeout option, in order to make
       // puppeteer’s timeout _never_ happen
-      let waitingForPageLoad = true
-      const loadPagePromise = page.goto(url, { timeout: timeout + 1000 })
-      if (pageLoadSkipTimeout) {
-        await Promise.race([
-          loadPagePromise,
-          new Promise(resolve => {
-            // instead we manually _abort_ page load after X time,
-            // in order to deal with spammy pages that keep sending non-critical requests
-            // (tracking etc), which would otherwise never load.
-            // With JS disabled it just shouldn't take that many seconds to load what's needed
-            // for critical viewport.
-            setTimeout(() => {
-              if (waitingForPageLoad) {
-                debuglog(
-                  'page load waiting ABORTED after ' +
-                    pageLoadSkipTimeout / 1000 +
-                    's. '
-                )
-                resolve()
-              }
-            }, pageLoadSkipTimeout)
+      const loadPageResponse = new Promise((resolve, reject) => {
+        page
+          .goto(url, {
+            timeout: timeout + 1000,
+            waitUntil: 'networkidle'
+          })
+          .then(response => {
+            // Reject exits the Promise.all call immediately and ensures that the pageload is always higher prio than timout
+            _hasLoaded = true
+            return resolve('loadPageResponse')
           })
+          .catch(err => {
+            _hasError = err
+            // Reject when error because we don't want an errored page
+            console.error(err)
+            return reject(err)
+          })
+      })
+
+      try {
+        // Instead of race we want to use all for better workflow. Race is evil
```

Comment:
this comment doesn't match the code, did you change your mind or is it a mistake? I am interested in why you think 'Race is evil'? To the best of my knowledge so far, Promise.race is a good fit here.

Reply:
The comment is just wrong. I changed it to Promise.all to be more safe. To explain:
The evil thing on race is, if you reject a promise the race is over and the promise race is rejected at all. But what we want to achieve is that the first promise resolve will win. Nevertheless I needed to change the workflow of promise.all to be reversed (reject is win). So I decided it is also not the best way. I just changed my mind. yeah